### PR TITLE
feat: BERT on SQuAD Dataset

### PR DESCRIPTION
**Pull request by naren-determined**
## Description
Example of fine-tuning BERT on SQuAD Dataset (compatible with both SQuAD1.1 and SQuAD2.0)  from https://github.com/huggingface/transformers/tree/master/examples/question-answering.

## Test Plan
Tested in both single GPU and distributed training contexts on AWS p2.8xlarge instance, which consists of 8 K80 GPUs and 488 GiB of RAM. Expected F1 score of 88.52 on bert-base model upon completion.

## Commentary (optional)

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: examples/experimental/bert_squad_pytorch/model_def.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        inputs = {
+            "input_ids": batch[0],
+            "attention_mask": batch[1],
+            "token_type_ids": batch[2],
+            "start_positions": batch[3],
+            "end_positions": batch[4],
+        }
+        outputs = model(**inputs)
+        loss = outputs[0]
+        self.tr_loss += loss.item()
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/model_def.py:120]: blocking: `loss` returned by `train_batch` should be the per-batch loss. we'll aggregate / average the loss for you. In other words, no need to track `self.tr_loss`, just return the `loss`.

File: examples/experimental/bert_squad_pytorch/data.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            max_seq_length=max_seq_length,
+            doc_stride=doc_stride,
+            max_query_length=max_query_length,
+            is_training=not evaluate,
+            return_dataset="pt",
+    )
+
+    if output_examples:
+        return dataset, examples, features
+    return dataset
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/data.py:30]: non-blocking (style): I know this code is copied directly huggingface, but imo it would be more readable to get rid of `output_examples` and always return the 3-tuple from this function. In situations where you don't need `examples` and `features` you can use: 

```
dataset, _, _ = load_and_cache_examples(...)
```

File: examples/experimental/bert_squad_pytorch/data.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+def load_and_cache_examples(data_dir: str, config, model_type, max_seq_length, doc_stride, max_query_length, model_name_or_path, evaluate=False, output_examples=False):
+    config_class, tokenizer_class, model_class = constants.MODEL_CLASSES[model_type]
+    tokenizer = tokenizer_class.from_pretrained(
+        config["model_name_or_path"], do_lower_case=True, cache_dir=None
+    )
+    try:
+        import tensorflow_datasets as tfds
+    except ImportError:
+        raise ImportError("Tensorflow_datasets needs to be installed.")
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/data.py:15]: blocking: I don't think this `try`, `except` block makes much sense in a Determined example, since we're explicitly calling out the dependencies in `startup-hook.sh`. I would move remove the try-except block and move this import with the other imports.

File: examples/experimental/bert_squad_pytorch/model_def.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                    p for n, p in model.named_parameters() if not any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": self.context.get_hparam("weight_decay"),
+            },
+            {
+                "params": [
+                    p for n, p in model.named_parameters() if any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": self.context.get_hparam("weight_decay"),
+            },
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/model_def.py:93]: blocking: I'm not sure I understand why this first argument to `AdamW` needs two items in the list. 

I looked at the original `run_squad` example also don't understand it, although it looks as though it uses different values for the `weight_decay` key (first using an arg, then using 0.0). Can you investigate what's going on here?

File: examples/experimental/bert_squad_pytorch/model_def.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        loss = outputs[0]
+        self.tr_loss += loss.item()
+        return {"loss": self.tr_loss}
+
+    def evaluate_full_dataset(self, data_loader: DataLoader, model: nn.Module) -> Dict[str, Any]:
+        tokenizer = self.tokenizer_class.from_pretrained(
+            self.context.get_data_config().get("model_name_or_path"),
+            do_lower_case=True,
+            cache_dir=None
+        )
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/model_def.py:128]: blocking: since you can reuse the tokenizer across multiple validation steps, you could move it to the trial instance constructor as `self.tokenizer`.

Comment (yoavz) [reply examples/experimental/bert_squad_pytorch/model_def.py:128]: actually, it looks like the same tokenizer is loaded in `load_and_cache_examples`, could you merge these paths to use the same tokenizer object? Probably only need to load it once at the beginning of the constructor.

File: examples/experimental/bert_squad_pytorch/model_def.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                eval_feature = self.features[feature_index.item()]
+                unique_id = int(eval_feature.unique_id)
+                output = [output[i].detach().cpu().tolist() for output in outputs]
+                start_logits, end_logits = output
+                result = SquadResult(unique_id, start_logits, end_logits)
+            all_results.append(result)
+
+        output_prediction_file = os.path.join(".", "predictions_{}.json".format(""))
+        output_nbest_file = os.path.join(".", "nbest_predictions_{}.json".format(""))
+        output_null_log_odds_file = None
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/model_def.py:149]: question: could we pass `None` instead of these filepaths to the transformers library `compute_predictions_logits` to prevent it from saving stuff to these paths? We're not using the result of these files in a Determined context, so better to avoid writing redundant stuff to disk.

File: examples/experimental/bert_squad_pytorch/model_def.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+
+class BertSQuADPyTorch(PyTorchTrial):
+    def __init__(self, context: det.TrialContext) -> None:
+        self.context = context
+        self.download_directory = f"/tmp/data-rank{self.context.distributed.get_rank()}"
+        self.config_class, self.tokenizer_class, self.model_class = constants.MODEL_CLASSES[
+            self.context.get_hparam("model_type")
+        ]
+        self.validation_dataset, self.examples, self.features = data.load_and_cache_examples(
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/model_def.py:31]: non-blocking: `self.examples` -> `self.validation_examples` and `self.features` -> `self.validation_features` -- don't want somebody mixing these attributes up for the training examples / features.

File: examples/experimental/bert_squad_pytorch/model_def.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+            config=self.context.get_data_config(),
+            model_type=self.context.get_hparam("model_type"),
+            max_seq_length=self.context.get_hparam("max_seq_length"),
+            doc_stride=self.context.get_hparam("doc_stride"),
+            max_query_length=self.context.get_hparam("max_query_length"),
+            model_name_or_path=self.context.get_data_config().get("model_name_or_path"),
+            evaluate=True,
+            output_examples=True,
+        )
+        self.tr_loss = 0.0
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/model_def.py:42]: blocking: see below, you don't need this .

File: examples/experimental/bert_squad_pytorch/model_def.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+            }
+            feature_indices = batch[3]
+            outputs = model(**inputs)
+            for i, feature_index in enumerate(feature_indices):
+                eval_feature = self.features[feature_index.item()]
+                unique_id = int(eval_feature.unique_id)
+                output = [output[i].detach().cpu().tolist() for output in outputs]
+                start_logits, end_logits = output
+                result = SquadResult(unique_id, start_logits, end_logits)
+            all_results.append(result)
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/model_def.py:145]: question: should line 145 be inside the for loop? you're only appending the `result` from the final feature index, not sure if this is intentional.

**Comment by yoavz** [comment]
@naren-determined let's discuss testing in person sometime when we meet next -- there are a few different options for how to test this example and where we put the test will depend on how lightweight the example is (data download time, required cpu / ram).

**Comment by yoavz** [comment]
blocking: use `git commit --amend` to make your commit descriptive according to https://www.conventionalcommits.org/en/v1.0.0

File: examples/experimental/bert_squad_pytorch/model_def.py (hunk 4 above)
Comment (naren-determined) [reply examples/experimental/bert_squad_pytorch/model_def.py:93]: Looking at link pointed to by this: https://github.com/huggingface/transformers/issues/492. In Lines 130-150 in the file pointed to by the link above, they seem to be modifying the OG Adam optimizer to handle L2 regularization in a particular way (while maintaining the original functionality with no weight decay). Essentially, they seem to be separating the weight decay out per parameter. So I believe in my code, Line 92 here should be setting `"weight_decay": 0.0` per what is in `run_squad`

File: examples/experimental/bert_squad_pytorch/const.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    max_seq_length: 384
+    doc_stride: 128
+    max_query_length: 64
+    n_best_size: 20
+    max_answer_length: 30
+    null_score_diff_threshold: 0.0
+searcher:
+    name: single
+    metric: f1
+    max_steps: 2
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/const.yaml:20]: blocking: let's set a more realistic amount of steps that converges to a reasonable f1 (whatever that is on this problem). add a comment at the top of the config calling out what metrics a user can expect.

File: examples/experimental/bert_squad_pytorch/data.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    with urllib.request.urlopen(validation_url) as url:
+        validation_data = json.loads(url.read().decode())
+        with open(validation_file, 'w') as f:
+            json.dump(validation_data, f)
+
+    if evaluate:
+        examples = processor.get_dev_examples(".", filename=validation_file)
+    else:
+        examples = processor.get_train_examples(".", filename=train_file)
+
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/data.py:34]: blocking: since `train_file` and `validation_file` are only used as an intermediate store for `processor.get_*_examples, consider using https://docs.python.org/3.7/library/tempfile.html#tempfile.NamedTemporaryFile instead. I think you could simplify this block something like: 

```
with train_file as tempfile.NamedTemporaryFile(), valid_file as tempfile.NamedTemporaryFile():
    with urllib.request.urlopen(train_url) as url:
        train_file.write(url.read().decode())
    
     examples = processor.get_dev_examples(".", filename=train_file.name)
```

It also has the added benefit that the `train_file` and `valid_file` won't be created on the filesystem as side-effects of this function since `tempfile` will clean up the files for you.

File: examples/experimental/bert_squad_pytorch/data.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+        train_url = "https://rajpurkar.github.io/SQuAD-explorer/dataset/train-v2.0.json"
+        validation_url = "https://rajpurkar.github.io/SQuAD-explorer/dataset/dev-v2.0.json"
+        train_file = "train-v2.0.json"
+        validation_file = "dev-v2.0.json"
+        processor = SquadV2Processor()
+
+    with urllib.request.urlopen(train_url) as url:
+        train_data = json.loads(url.read().decode())
+        with open(train_file, 'w') as f:
+            json.dump(train_data, f)
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/data.py:24]: blocking: json load and dump roundtrip is redundant, I think you can directly do `f.write(url.read().decode())`

File: examples/experimental/bert_squad_pytorch/const.yaml [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    smaller_is_better: false
+min_validation_period: 200 
+data:
+    model_name_or_path: "bert-base-uncased"
```
Comment (yoavz) [inline examples/experimental/bert_squad_pytorch/const.yaml:24]: I have two questions based on the naming `model_name_or_path`.

1) is there a scenario in which this would be a `path` instead of a model name, in the context of this Determined example? If not, I would just call it `model_name`.
2) is this used for figuring out how to initialize the weights of this training job? if so, a better name might make that explicit: `pretrained_model_name`.

Comment (naren-determined) [reply examples/experimental/bert_squad_pytorch/const.yaml:24]: Good point here
